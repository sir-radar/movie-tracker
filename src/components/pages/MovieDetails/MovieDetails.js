import React,{useEffect, useState} from 'react';
import { connect } from 'react-redux';
import YouTube from 'react-youtube';
import { animateScroll } from "react-scroll";
import Nav from '../../Nav/Nav';
import Loader from '../../Loader/Loader';
import MoviePoster from '../../MoviePoster/MoviePoster';
import { getMovieDetails } from '../../../store/actions/movieActions';

const MovieDetails = ({movieStatus, match, getMovieDetails, movieDetails }) => {

  useEffect(() => {
    getMovieDetails(match.params.id)
  },[getMovieDetails, match.params.id]);

  const [showTrialer, setShowTrailer] = useState(false);

  const displayTrailer = (e) =>{
    e.preventDefault();
    setShowTrailer(true)
    //scrolls to the bottom of page to show trailer
    animateScroll.scrollToBottom()
  }

  //youtube options
  const opts = {
    height: '390',
    width: '450',
    playerVars: {
      autoplay: 1
    }
  };

  return (
    <>
      <Nav/>
      <main data-testid="movie-details-page" className="mt-4 p-3 col-10 col-md-10 offset-0 offset-md-1">
        <div className="row">

          {
            (movieStatus === "PENDING") ? <Loader/> : null
          }

          {
            (movieStatus === "SUCCESS") ? <MoviePoster image={movieDetails.poster_path}/> : null
          }


          {
            (movieStatus === "SUCCESS") ? 
              <div className="col-12 col-md-8 mb">
                <div className="col-12">
                  <h2>{movieDetails.title}</h2>
                  <p><span>Genres:</span> <span>{ movieDetails.genres.map(genre => genre.name) }</span></p>
                  <p><span>Average vote:</span> <span>{movieDetails.vote_average}</span></p>
                  <p><span>Language(s):</span><span>{movieDetails.spoken_languages.map(lang => lang.name)}</span></p>
                  <p>{movieDetails.overview}</p>
                  <p><span>Realease:</span><span>{movieDetails.release_date}</span></p>
                  <p><span>Duration:</span><span>{movieDetails.runtime} minutes</span></p>

                  {
                    (!showTrialer && movieDetails.videos.results.length > 0) ? 
                          <button onClick={(e) => displayTrailer(e)} className="btn btn-primary">Watch Trailer</button>
                          : null
                  }
                  
                </div>
              </div>

            : null
          }
        </div>
        <div className="col-12 col-md-6 mt-4 mx-0 mx-md-auto">
            {
              (showTrialer) ? <YouTube
                                videoId={movieDetails.videos.results[0].key}
                                opts={opts}
                          />
                : null
            }
        </div>
      </main>
    </>
  );
}

const mapStateToProps = (state) => (
  {
    movieStatus: state.status.movieDetailsAction,
    movieDetails: state.movie,
  }
);

const mapDispatchToProps = (dispatch) => (
  {
    getMovieDetails: (movieId) => dispatch(getMovieDetails(movieId)),
  }
)

export default connect(mapStateToProps, mapDispatchToProps) (MovieDetails);
