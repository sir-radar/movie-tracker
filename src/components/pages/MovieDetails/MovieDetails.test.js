import React from 'react';
import { MemoryRouter } from "react-router-dom";
import { render, cleanup, fireEvent } from '@testing-library/react';
import MovieDetails from './MovieDetails';
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { Provider } from 'react-redux';
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";


const buildStore = configureStore([thunk]);

afterEach(cleanup)

const renderComponent = (states) => {
  //mock store
  const store = buildStore({...states});
  const div = document.createElement("div");
  const props={
    params:{
      id: 4566
    }
  }
  return render(<Provider store={store}><MemoryRouter><MovieDetails match={props} /></MemoryRouter></Provider>, div)
}

const renderComponentForSnapshot = (states) => {
  const store = buildStore({...states});
  const props={
    params:{
      id: 4566
    }
  }
  return <Provider store={store}><MemoryRouter><MovieDetails match={props} /></MemoryRouter></Provider>
}

describe('MovieDetails Page Component', () => {

  it("renders without crashing", async () => {

    const states = {
      status:{
        movieDetailsAction: 'PENDING'
      },
      movie:{},
    }

    renderComponent(states)
    
  })

  it("renders nav component", () => {

    const states = {
      status:{
        movieDetailsAction: ''
      },
      movie:{},
    }

    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("nav")).not.toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("movie-poster")).toBeNull()
    expect(queryByTestId("movie-summary")).toBeNull()
    expect(queryByTestId("trailer")).toBeNull()
  })

  it("renders loading component on pending request", () => {

    const states = {
      status:{
        movieDetailsAction: 'PENDING'
      },
      movie:{}
    }

    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("nav")).not.toBeNull()
    expect(queryByTestId("loader")).not.toBeNull()
    expect(queryByTestId("movie-poster")).toBeNull()
    expect(queryByTestId("movie-summary")).toBeNull()
    expect(queryByTestId("trailer")).toBeNull()
  })

  it("renders Movie Poster component and summary on successful request", () => {

    const states = {
      status:{
        movieDetailsAction: 'SUCCESS'
      },
      movie:{
        poster_path: "",
        genres:[],
        spoken_languages:[],
        videos:{
          results:[]
        }
      }
    }

    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("nav")).not.toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("movie-poster")).not.toBeNull()
    expect(queryByTestId("movie-summary")).not.toBeNull()
    expect(queryByTestId("trailer")).toBeNull()

  })


  it("shows trailer when show trailer button is clicked", () => {
    const states = {
      status:{
        movieDetailsAction: 'SUCCESS'
      },
      movie:{
        poster_path: "",
        genres:[],
        spoken_languages:[],
        videos:{
          results:[
            {key:"TjjUklNKLL"}
          ]
        }
      }
    }

    const {queryByTestId, getByTestId} = renderComponent(states)
    expect(queryByTestId("nav")).not.toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("movie-poster")).not.toBeNull()
    expect(queryByTestId("movie-summary")).not.toBeNull()
    expect(queryByTestId("trailer-trigger")).not.toBeNull()
    fireEvent.click(getByTestId('trailer-trigger'))
    expect(queryByTestId("trailer")).not.toBeNull()

  })

  
  it("matches snapshot", () => {
    const states = {
      status:{
        movieDetailsAction: 'SUCCESS'
      },
      movie:{
        poster_path: "",
        genres:[],
        spoken_languages:[],
        videos:{
          results:[
            {key:"TjjUklNKLL"}
          ]
        }
      }
    }
    const tree = renderer.create(renderComponentForSnapshot(states)).toJSON();
    expect(tree).toMatchSnapshot();
  })

})