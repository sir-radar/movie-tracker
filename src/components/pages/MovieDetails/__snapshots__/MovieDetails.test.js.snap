// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Page Component matches snapshot 1`] = `
Array [
  <nav
    className="d-flex justify-content-center"
    data-testid="nav"
  >
    <a
      className="p-3 primary-text-color active"
      data-testid="search"
      href="/"
      onClick={[Function]}
    >
      Search
    </a>
    <a
      className="p-3 primary-text-color"
      data-testid="favourite"
      href="/favourites"
      onClick={[Function]}
    >
      Favorites
    </a>
    <a
      className="p-3 primary-text-color"
      data-testid="watchlater"
      href="/watchlist"
      onClick={[Function]}
    >
      WatchLater
    </a>
  </nav>,
  <main
    className="mt-4 p-3 col-10 col-md-10 offset-0 offset-md-1"
    data-testid="movie-details-page"
  >
    <div
      className="row"
    >
      <div
        className="col-12 col-md-4"
        data-testid="movie-poster"
      >
        <div
          className="col-12"
        >
          <img
            alt="..."
            className="h-25 w-100"
            src="/no-image.png"
          />
        </div>
      </div>
      <div
        className="col-12 col-md-8 mb"
        data-testid="movie-summary"
      >
        <div
          className="col-12"
        >
          <h2 />
          <p>
            <span>
              Genres: 
            </span>
             
            <span />
          </p>
          <p>
            <span>
              Average vote: 
            </span>
             
            <span />
          </p>
          <p>
            <span>
              Language(s): 
            </span>
            <span />
          </p>
          <p />
          <p>
            <span>
              Realease: 
            </span>
            <span />
          </p>
          <p>
            <span>
              Duration: 
            </span>
            <span>
               minutes
            </span>
          </p>
          <button
            className="btn btn-primary"
            data-testid="trailer-trigger"
            onClick={[Function]}
          >
            Watch Trailer
          </button>
        </div>
      </div>
    </div>
  </main>,
]
`;
